### PR TITLE
get username from data dict if not available in context dict of event

### PR DESCRIPTION
**Pull request by rehanedly**
**Description:** Describe in a couple of sentences what this PR adds

**JIRA:** Link to JIRA ticket

**Dependencies:** dependencies on other outstanding PRs, issues, etc.

**Merge deadline:** List merge deadline (if any)

**Installation instructions:** List any non-trivial installation
instructions.

**Testing instructions:**

1. Open page A
2. Do thing B
3. Expect C to happen
4. If D happened instead - check failed.

**Merge checklist:**
- [ ] All reviewers approved
- [ ] CI build is green
- [ ] Version bumped
- [ ] Changelog record added
- [ ] Documentation updated (not only docstrings)
- [ ] Commits are squashed

**Post merge:**
- [ ] Create a tag
- [ ] Check new version is pushed to PyPI after tag-triggered build is
      finished.

**Author concerns:** List any concerns about this PR - inelegant
solutions, hacks, quick-and-dirty implementations, concerns about
migrations, etc.

### CODE REVIEW

**Comment by edx-webhook** [comment]
Thanks for the pull request, @rehanedly! I've created [OSPR-5824](https://openedx.atlassian.net/browse/OSPR-5824) to keep track of it in JIRA, where we prioritize reviews. Please note that it may take us up to several weeks or months to complete a review and merge your PR.

Feel free to add as much of the following information to the ticket:
- supporting documentation
- [Open edX discussion forum threads](https://discuss.openedx.org/)
- timeline information ("this must be merged by XX date", and why that is)
- partner information ("this is a course on edx.org")
- any other information that can help Product understand the context for the PR

All technical communication about the code itself will be done via the GitHub pull request interface. As a reminder, [our process documentation is here](http://edx-developer-guide.readthedocs.org/en/latest/process/overview.html).


Please let us know once your PR is ready for our review and all tests are green.

**Review by ziafazal — COMMENTED** [review]
@rehanedly I think we should add a method in BaseTransformerMixin to extract username something like `def extract_username(self)`

File: event_routing_backends/processors/tests/fixtures/current/edx.course.enrollment.activated.enterprise.json [hunk 1]
```diff
@@ -0,0 +1,24 @@
+{
+    "context": {
+        "accept_language": "en-US,en;q=0.9",
+        "agent": "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_4) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/83.0.4103.116 Safari/537.36",
+        "client_id": null,
+        "course_id": "course-v1:edX+DemoX+Demo_Course",
+        "host": "localhost:18000",
+        "ip": "172.18.0.1",
+        "org_id": "edX",
```
Comment (ziafazal) [inline event_routing_backends/processors/tests/fixtures/current/edx.course.enrollment.activated.enterprise.json:9]: we should pass in `enterprise_uuid` here to cater for the enterprise use case and in in expected data we should assert it in context.

File: event_routing_backends/processors/mixins/base_transformer.py [hunk 2]
```diff
@@ -102,3 +102,16 @@ def transform(self):
                 )
 
         return self.transformed_event
+
+    def extract_username(self):
+        """
+        extract username edX event.
```
Comment (ziafazal) [inline event_routing_backends/processors/mixins/base_transformer.py:108]: `Extracts username from event by finding it in context first and falling back to data if context does have username key`

File: event_routing_backends/processors/mixins/base_transformer.py [hunk 3]
```diff
@@ -102,3 +102,16 @@ def transform(self):
...
+
+        Returns:
+            username
```
Comment (ziafazal) [inline event_routing_backends/processors/mixins/base_transformer.py:111]: I think it should type of return value like `str`

File: event_routing_backends/processors/mixins/base_transformer.py [hunk 4]
```diff
@@ -102,3 +102,16 @@ def transform(self):
...
+        """
+        username = self.event['context']['username']
```
Comment (ziafazal) [inline event_routing_backends/processors/mixins/base_transformer.py:113]: can we simplify these statement to something like this?
`return self.event['context'].get('username', self.event['data'].get('username'))`

File: event_routing_backends/processors/xapi/tests/fixtures/expected/edx.course.enrollment.activated.enterprise.json [hunk 5]
```diff
@@ -0,0 +1,27 @@
+{
+    "actor": {
```
Comment (ziafazal) [inline event_routing_backends/processors/xapi/tests/fixtures/expected/edx.course.enrollment.activated.enterprise.json:2]: we need to add a test for caliper too

**Review by ziafazal — CHANGES_REQUESTED** [review]
@rehanedly I have few suggestions. Could you please take a look?

File: event_routing_backends/processors/mixins/base_transformer.py (hunk 4 above)
Comment (rehanedly) [reply event_routing_backends/processors/mixins/base_transformer.py:113]: 'str' object has no attribute 'get'

File: event_routing_backends/processors/tests/fixtures/current/edx.course.enrollment.activated.enterprise.json (hunk 1 above)
Comment (rehanedly) [reply event_routing_backends/processors/tests/fixtures/current/edx.course.enrollment.activated.enterprise.json:9]: done

File: event_routing_backends/processors/xapi/tests/fixtures/expected/edx.course.enrollment.activated.enterprise.json (hunk 5 above)
Comment (rehanedly) [reply event_routing_backends/processors/xapi/tests/fixtures/expected/edx.course.enrollment.activated.enterprise.json:2]: done

File: event_routing_backends/processors/mixins/base_transformer.py (hunk 3 above)
Comment (rehanedly) [reply event_routing_backends/processors/mixins/base_transformer.py]: done

File: event_routing_backends/processors/mixins/base_transformer.py (hunk 4 above)
Comment (rehanedly) [reply event_routing_backends/processors/mixins/base_transformer.py]: self.event['context'].get('username', self.event['data'].get('username'))
This line of code will return username in data dict if username key is not set in context dict. But in our case username key is set in the context dict but it's value is empty string ''. so we need to have a check if user name in context dic is empty as well as it exist in data dict. So I have added above code that is working fine

**Comment by natabene** [comment]
@rehanedly Thank you for your contribution. @jmbowman FYI this OSPR is coming.

**Comment by natabene** [comment]
@ziafazal Can you let me know once it is ready for edX review?

File: event_routing_backends/processors/mixins/base_transformer.py [hunk 6]
```diff
@@ -102,3 +102,16 @@ def transform(self):
...
+    def extract_username(self):
+        """
+        Extracts username from event by finding it in context first and falling back to data
+        if context does have username key
+
+        Returns:
+            str
+        """
+
+        return self.event['data']['username'] \
```
Comment (ziafazal) [inline event_routing_backends/processors/mixins/base_transformer.py:115]: nit: I think in order to have protection against `KeyError` we do something like this
```
username = self.event.get('context', {}).get('username')
if not username:
    username = self.event.get('data', {}).get('username')
return username
```

Comment (rehanedly) [reply event_routing_backends/processors/mixins/base_transformer.py]: done

**Review by ziafazal — APPROVED** [review]
👍

**Comment by ziafazal** [comment]
@natabene it is now ready for edX review.

**Comment by nasthagiri** [comment]
@ziafazal Since you are a core committer of this repo and have not authored this PR yourself, you do not need edX review.  See the [guidelines for core committer reviews](https://openedx.atlassian.net/wiki/spaces/COMM/pages/2718564358/Core%2BCommitters%2BInitial%2BPhases%2Bfor%2BPilot%2BRollout#Reviews-and-Ownership).

**Comment by ziafazal** [comment]
@nasthagiri Thanks for clarifying. We previously merged PRs after my review I thought may be something has changed now.

**Comment by edx-webhook** [comment]
@rehanedly 🎉 Your pull request was merged!

Please take a moment to [answer a two question survey](https://docs.google.com/forms/d/e/1FAIpQLSceJOyGJ6JOzfy6lyR3T7EW_71OWUnNQXp68Fymsk3MkNoSDg/viewform?usp=pp_url&entry.1671973413=edx/event-routing-backends&entry.867055334=https://github.com/edx/event-routing-backends/pull/78&entry.1484655318=https://github.com/rehanedly&entry.752974735=2021-06-02+07:54&entry.1917517419=2021-06-07+07:03&entry.2133058324=Yes) so we can improve your experience in the future.

**Pull request merged**